src/AggregateFunctions/AggregateFunctionDeepMergeJSON.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include <map>
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <Columns/ColumnObject.h>
+#include <DataTypes/DataTypeObject.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <base/StringRef.h>
+#include <Common/Arena.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+/// Reasonable limits to prevent memory abuse
+constexpr size_t MAX_JSON_MERGE_PATHS = 10000;
+constexpr size_t MAX_JSON_MERGE_PATH_LENGTH = 1000;
+constexpr size_t MAX_JSON_MERGE_TOTAL_SIZE = 100_MiB;
+
+/// JSON unset marker key
+constexpr const char * UNSET_KEY = "$unset";
+
+struct DeepMergeJSONAggregateData
+{
+    struct PathData
+    {
+        Field value;
+        size_t row_order;
+        // Track if this path was explicitly deleted
+        bool is_deleted = false;
+    };
+
+    /// Use std::map to keep paths sorted for consistent output
+    /// StringRef will point to Arena-allocated memory
+    std::map<StringRef, PathData> paths;
+    size_t row_count = 0;
```

Comment:
I think this and `row_order` are useless, as they don't guarantee anything. We can just always keep the latest value added to the aggregate data. And during merging of 2 aggregate data we can treat the value fron the right data as the latest